src/main/java/com/google/firebase/database/connection/WebsocketConnection.java

```diff
@@ -108,78 +99,86 @@ public void run() {
     }
     isClosed = true;
-    // Although true is passed for both of these, they each run on the same event loop, so
-    // they will
-    // never be running.
     conn.close();
+
+    // Although true is passed for both of these, they each run on the same event loop, so
+    // they will never be running.
     if (connectTimeout != null) {
       connectTimeout.cancel(true);
+      connectTimeout = null;
     }
     if (keepAlive != null) {
       keepAlive.cancel(true);
       keepAlive = null;
     }
   }
 
-  public void send(Map<String, Object> message) {
+  void send(Map<String, Object> message) {
     resetKeepAlive();
-
     try {
       String toSend = JsonMapper.serializeJson(message);
-      String[] segs = splitIntoFrames(toSend, MAX_FRAME_SIZE);
-      if (segs.length > 1) {
-        conn.send("" + segs.length);
+      List<String> frames = splitIntoFrames(toSend, MAX_FRAME_SIZE);
+      if (frames.size() > 1) {
+        conn.send("" + frames.size());
       }
 
-      for (String seg : segs) {
+      for (String seg : frames) {
         conn.send(seg);
       }
     } catch (IOException e) {
       logger.error("Failed to serialize message: " + message.toString(), e);
-      shutdown();
+      closeAndNotify();
     }
   }
 
-  private void appendFrame(String message) {
-    frameReader.addString(message);
-    totalFrames -= 1;
-    if (totalFrames == 0) {
-      // Decode JSON
-      try {
-        frameReader.freeze();
-        Map<String, Object> decoded = JsonMapper.parseJson(frameReader.toString());
-        if (logger.logsDebug()) {
-          logger.debug("handleIncomingFrame complete frame: " + decoded);
-        }
-        delegate.onMessage(decoded);
-      } catch (IOException e) {
-        logger.error("Error parsing frame: " + frameReader.toString(), e);
-        close();
-        shutdown();
-      } catch (ClassCastException e) {
-        logger.error("Error parsing frame (cast error): " + frameReader.toString(), e);
-        close();
-        shutdown();
-      } finally {
-        frameReader = null;
+  private List<String> splitIntoFrames(String src, int maxFrameSize) {
+    if (src.length() <= maxFrameSize) {
+      return ImmutableList.of(src);
+    } else {
+      ImmutableList.Builder<String> frames = ImmutableList.builder();
+      for (int i = 0; i < src.length(); i += maxFrameSize) {
+        int end = Math.min(i + maxFrameSize, src.length());
+        String seg = src.substring(i, end);
+        frames.add(seg);
       }
+      return frames.build();
     }
   }
 
   private void handleNewFrameCount(int numFrames) {
-    totalFrames = numFrames;
-    frameReader = new StringListReader();
     if (logger.logsDebug()) {
-      logger.debug("HandleNewFrameCount: " + totalFrames);
+      logger.debug("HandleNewFrameCount: " + numFrames);
+    }
+    buffer.initialize(numFrames);
+  }
+
+  private void appendFrame(String message) {
+    if (buffer.append(message) > 0) {
+      return;
```

Comment:
it wasn't obvious that buffer.append() would return the number of remaining frames, so maybe add a comment:
```
return; // waiting for more frames
```

or use an intermediate variable:

```
int framesRemaining = buffer.append(message);
if (framesRemaining > 0) {
  return;
}
```

or ignore the return value:

```
buffer.append(message);
if (buffer.hasRemaining()) {
  return;
}
```

Reply:
Done. Used option 2.